### PR TITLE
refactor(docs-infra): prepare DocumentService to handle new disambiguated URLs

### PR DESCRIPTION
**Pull request by petebacondarwin**
A subsequent commit is going to change disambiguated URLs.
This commit prepares the AIO application to attempt the new URLs
if the old URLs fail. This will help to mitigate problems that may occur
during the period between deployment of the new version and the
service-worker not being updated.

### CODE REVIEW

File: aio/src/app/documents/document.service.ts [hunk 1]
```diff
@@ -66,7 +66,22 @@ export class DocumentService {
...
+        }),
+        catchError((error: HttpErrorResponse) => {
+          const disambiguatedPath = convertDisambiguatedPath(requestPath);
+          return error.status === 404 && disambiguatedPath !== requestPath ?
+              this.http.get<DocumentContents>(convertDisambiguatedPath(requestPath)) :
+              throwError(error);
+        }),
+        // END HACK: PREPARE FOR CHANGING TO CASE-INSENSITIVE URLS
+        catchError((error: HttpErrorResponse) => {
+          console.log('3');
```
Comment (gkalpak) [inline aio/src/app/documents/document.service.ts:84]:
```suggestion
```

:grin:

File: aio/src/app/documents/document.service.ts [hunk 2]
```diff
@@ -66,7 +66,22 @@ export class DocumentService {
             throw Error('Invalid data');
           }
         }),
+        // HACK: PREPARE FOR CHANGING TO CASE-INSENSITIVE URLS
         catchError((error: HttpErrorResponse) => {
+          const encodedPath = encodeToLowercase(requestPath);
```
Comment (gkalpak) [inline aio/src/app/documents/document.service.ts:71]: Why is this needed? Can't we go directly to `convertDisambiguatedPath()`?

**Review by gkalpak — CHANGES_REQUESTED** [review]
The general concept lgtm :+1:
Just a couple of minor comments/questions.

File: aio/src/app/documents/document.service.ts (hunk 2 above)
Comment (petebacondarwin) [reply aio/src/app/documents/document.service.ts:71]: I was trying to be conservative. For example what if there really was a file that was `Xyz-1` (i.e. the `-1` was actually really part of the path and not a disambiguator)? Then the second step would only try `x_yz.json`, whereas actually we would want `x_yz-1.json`.

(I appreciated that I don't think we currently have such a doc but I thought this would just be safest).

File: aio/src/app/documents/document.service.ts [hunk 3]
```diff
@@ -66,7 +66,22 @@ export class DocumentService {
...
+          return error.status === 404 && encodedPath !== requestPath ?
+              this.http.get<DocumentContents>(encodeToLowercase(requestPath)) :
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/src/app/documents/document.service.ts:73]: 🤦

File: aio/src/app/documents/document.service.ts [hunk 4]
```diff
@@ -66,7 +66,22 @@ export class DocumentService {
...
+              throwError(error);
+        }),
+        catchError((error: HttpErrorResponse) => {
+          const disambiguatedPath = convertDisambiguatedPath(requestPath);
+          return error.status === 404 && disambiguatedPath !== requestPath ?
+              this.http.get<DocumentContents>(convertDisambiguatedPath(requestPath)) :
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/src/app/documents/document.service.ts:79]: 🤦

File: aio/src/app/documents/document.service.ts (hunk 1 above)
Comment (petebacondarwin) [reply aio/src/app/documents/document.service.ts:84]: 😱

**Comment by petebacondarwin** [comment]
You can test this functionality by going to https://pr42509-0e89f2e.ngbuilds.io/docs/X_A-1 and checking out the network requests:

Screenshot 2021-06-07 at 20 13 13

**Review by gkalpak — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking

**Review by atscott — APPROVED** [review]
reviewed-for: size-tracking
